File: 60.Permutation Sequence/src/Solution.java
public class Solution {
    public String getPermutation(int n, int k) {
        String ll = "";
    	for (int i = 1; i < n+1; i++) {
        	ll += Integer.toString(i);
        }
    	int divisor = 1;
    	for (int i = 1; i < n; i++) {
    		divisor *= i;
    	}
    	String answer = "";
    	
    	int group_num = 0;
    	char choose;
    	String ll2 = "";
    	while (k > 0 && k <= divisor*n) {
    		group_num = k / divisor;
    		k %= divisor;
    		
    		if (k > 0) {
    			choose = ll.charAt(group_num);
    			ll = ll.substring(0, group_num) + ll.substring(group_num+1);
    			answer += choose;
    		}
    		else {
    			choose = ll.charAt(group_num - 1);
    			ll = ll.substring(0, group_num-1) + ll.substring(group_num);
    			answer += choose;
    			StringBuffer sb=new StringBuffer(ll);
    			sb.reverse();
    			ll2 = sb.toString();
    			answer = answer + ll2;
    			break;
    		}
    		divisor /= ll.length();
    	}
    	return answer;
    }
}

File: 60.Permutation Sequence/Solution.py
class Solution(object):
    # def getPermutation(self, n, k):
    #     """
    #     :type n: int
    #     :type k: int
    #     :rtype: str
    #     """
    #     result = []
    #     num = []
    #     string = ""
    #     for i in range(1, n+1):
    #         num.append(i)
    #     self.DFS(num, string, result, k)
    #     # print len(result)
    #     return result.pop()
    #
    #
    # def DFS(self, num, string, result, k):
    #     if len(result) == k:
    #         return
    #     if len(num) == 0:
    #         result.append(string)
    #     else:
    #         for i in range(len(num)):
    #             self.DFS(num[:i] + num[i+1:], string + str(num[i]), result, k)


    def getPermutation(self, n, k):
        ll = [str(i) for i in range(1, n+1)]
        divisor = 1
        for i in range(1, n):
            divisor *= i
        answer = ""

        while k > 0 and k <= divisor * n:

            group_num = k / divisor
            # print group_num
            k %= divisor
            # print k

            if k > 0:
                # print "SSS"
                choose = ll.pop(group_num)
                answer += choose
            else:
                # print "AAA"
                choose = ll.pop(group_num - 1)
                answer += choose
                ll.reverse()

                to_add = "".join(ll)
                answer += to_add
                break

            divisor /= len(ll)

        return answer


if __name__ == '__main__':
    S = Solution()
    print S.getPermutation(3,4)

